paddle/fluid/operators/put_along_axis_op.h

```diff
@@ -0,0 +1,115 @@
+    result->Resize(index->dims());
+    result->mutable_data<T>(ctx.GetPlace());
+
+    const platform::DeviceContext &device_ctx = ctx.device_context();
+    const auto &index_type = index->type();
+    if (reduce_op == "add") {
+      if (index_type == framework::proto::VarType::INT32) {
+        cpu_scatter_add_kernel<T, int32_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      } else if (index_type == framework::proto::VarType::INT64) {
+        cpu_scatter_add_kernel<T, int64_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      }
+    } else if (reduce_op == "multiply" || reduce_op == "mul") {
+      if (index_type == framework::proto::VarType::INT32) {
+        cpu_scatter_mul_kernel<T, int32_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      } else if (index_type == framework::proto::VarType::INT64) {
+        cpu_scatter_mul_kernel<T, int64_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      }
+    } else if (reduce_op == "assign") {
+      if (index_type == framework::proto::VarType::INT32) {
+        cpu_scatter_assign_kernel<T, int32_t>(*input, axis, *index, *value,
+                                              device_ctx);
+      } else if (index_type == framework::proto::VarType::INT64) {
+        cpu_scatter_assign_kernel<T, int64_t>(*input, axis, *index, *value,
+                                              device_ctx);
+      }
+    } else {
+      platform::errors::InvalidArgument(
+          "can not suppor reduce_op: (%s) for scatter kernel, only "
+          "support reduce op: 'add‘, 'assign', 'multiply', the defalut reduce "
+          "op is 'assign' ",
+          reduce_op);
+      return;
+    }
+    *result = *input;  // inplace opeartion
+  }
+};
+
+template <typename T>
+class PutAlongAxisGradOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    PADDLE_ENFORCE_EQ(
+        platform::is_cpu_place(ctx.GetPlace()), true,
+        platform::errors::PreconditionNotMet("This kernel only runs on CPU."));
+
+    auto input_grad = ctx.Output<Tensor>(framework::GradVarName("Input"));
+    auto index = ctx.Input<Tensor>("Index");
+    auto result_grad = ctx.Input<Tensor>(framework::GradVarName("Result"));
+    auto axis = ctx.Attr<int>("Axis");
+    // We need to know the shape of input matrix to determine the shape of grad
+    // matrix of input.
+    auto input = ctx.Input<Tensor>("Input");
+    input_grad->Resize(input->dims());
+    input_grad->mutable_data<T>(ctx.GetPlace());
+    auto input_grad_data = input_grad->data<T>();
+    for (int64_t i = 0; i < input_grad->numel(); ++i) {
+      *(input_grad_data + i) = 0;
+    }
+    const auto &index_type = index->type();
+    if (index_type == framework::proto::VarType::INT32) {
```

Comment:
I think there may be something wrong here.

Though put along axis is implemented as inplace operation. The gradient operation should compute the gradient of Input and Value separately as needed.

So not only is taking output's gradient along axis with index to fill Value's gradient needed, copying gradients of the output to the input's gradient among those positions that is not gathered and masking (for scatter assign) gradient of the input amopng those gathered positions is also needed.